src/bin/fd_filestat_set.rs

```diff
@@ -0,0 +1,107 @@
+use libc;
+use misc_tests::open_scratch_directory;
+use misc_tests::utils::{cleanup_file, close_fd};
+use misc_tests::wasi::{wasi_fd_filestat_set_size, wasi_fd_filestat_set_times, wasi_fd_filestat_get, wasi_path_open};
+use std::{env, process};
+
+fn test_file_allocate(dir_fd: libc::__wasi_fd_t) {
+    // Create a file in the scratch directory.
+    let mut file_fd = libc::__wasi_fd_t::max_value() - 1;
+    let status = wasi_path_open(
+        dir_fd,
+        0,
+        "file",
+        libc::__WASI_O_CREAT,
+        libc::__WASI_RIGHT_FD_READ | libc::__WASI_RIGHT_FD_WRITE,
+        0,
+        0,
+        &mut file_fd,
+    );
+    assert_eq!(status, libc::__WASI_ESUCCESS, "opening a file");
+    assert!(
+        file_fd > libc::STDERR_FILENO as libc::__wasi_fd_t,
+        "file descriptor range check",
+    );
+
+    // Check file size
+    let mut stat = libc::__wasi_filestat_t {
+        st_dev: 0,
+        st_ino: 0,
+        st_filetype: 0,
+        st_nlink: 0,
+        st_size: 0,
+        st_atim: 0,
+        st_mtim: 0,
+        st_ctim: 0,
+    };
+    let status = wasi_fd_filestat_get(file_fd, &mut stat);
+    assert_eq!(status, libc::__WASI_ESUCCESS, "reading file stats");
+    assert_eq!(stat.st_size, 0, "file size should be 0");
+
+    // Allocate some size
+    let status = wasi_fd_filestat_set_size(file_fd, 100);
+    assert_eq!(status, libc::__WASI_ESUCCESS, "fd_filestat_set_size");
+
+    let status = wasi_fd_filestat_get(file_fd, &mut stat);
+    assert_eq!(
+        status,
+        libc::__WASI_ESUCCESS,
+        "reading file stats after fd_filestat_set_size"
+    );
+    assert_eq!(stat.st_size, 100, "file size should be 100");
+
+    // Allocate should not modify if less than current size
+    let old_atim = stat.st_atim;
+    let new_mtim = stat.st_mtim - 100;
+    let status = wasi_fd_filestat_set_times(file_fd, new_mtim, new_mtim, libc::__WASI_FILESTAT_SET_MTIM);
+    assert_eq!(
+        status,
+        libc::__WASI_ESUCCESS,
+        "fd_filestat_set_times"
+    );
+
+    let status = wasi_fd_filestat_get(file_fd, &mut stat);
+    assert_eq!(
+        status,
+        libc::__WASI_ESUCCESS,
+        "reading file stats after fd_filestat_set_times"
+    );
+    assert_eq!(
+        stat.st_size, 100,
+        "file size should remain unchanged at 100"
+    );
+    assert_eq!(
+        stat.st_mtim, new_mtim, "mtim should change"
+    );
+    assert_eq!(
+        stat.st_atim, old_atim, "atim should not change"
```

Comment:
Would you mind swapping `new_mtim` and `old_atim` for expected literal values? I find that tests which use variables for testing, sometimes can produce misleading results. :-)

Reply:
I need to remember the old values of `stat.st_atim` and `stat.st_mtim` somewhere, because it's `stat` is reused between `fd_filestat_get` calls. I can alternatively clone the whole `stat` structure but I'm not sure if it's cleaner.

Reply:
Ah, of course, that's fair enough! :-)